client/my-sites/checkout/composite-checkout/lib/paypal-express-processor.ts

```diff
@@ -27,28 +27,32 @@ export default async function payPalProcessor(
 		siteId,
 		siteSlug,
 		contactDetails,
-		customizedCancelUrl,
 	} = transactionOptions;
 	reduxDispatch( recordTransactionBeginAnalytics( { paymentMethodId: 'paypal' } ) );
 
 	const thankYouUrl = getThankYouUrl();
 	let currentUrl;
 	let currentBaseUrl;
+	let currentUrlQuery;
 	try {
 		currentUrl = window.location.href;
 		currentBaseUrl = window.location.origin;
+		currentUrlQuery = window.location.search;
 	} catch ( error ) {
 		currentUrl = `https://wordpress.com/checkout/${ siteSlug }`;
 		currentBaseUrl = 'https://wordpress.com';
+		currentUrlQuery = '';
 	}
 	const currentUrlWithoutQuery = currentUrl.split( /\?|#/ )[ 0 ];
+	const updatedQuery = new URLSearchParams( currentUrlQuery );
 	const successUrl = thankYouUrl.startsWith( 'http' ) ? thankYouUrl : currentBaseUrl + thankYouUrl;
-	let cancelUrl = customizedCancelUrl;
-	if ( ! cancelUrl ) {
-		cancelUrl = createUserAndSiteBeforeTransaction
-			? currentUrlWithoutQuery + '?cart=no-user'
-			: currentUrlWithoutQuery;
+	if ( createUserAndSiteBeforeTransaction ) {
+		updatedQuery.append( 'cart', 'no-user' );
 	}
+	const updatedQueryString = updatedQuery.toString();
+	const cancelUrl = updatedQueryString.length
+		? currentUrlWithoutQuery + '?' + updatedQueryString
+		: currentUrlWithoutQuery;
```

Comment:
[NIT] Maybe we can simplify these codes by `new URL` as followed:
```js
let currentUrl;
try {
  currentUrl = new URL( window.location.href );
} catch ( error ) {
  currentUrl = new URL( `https://wordpress.com/checkout/${ siteSlug }` );
}

if ( createUserAndSiteBeforeTransaction ) {
  currentUrl.searchParams.set( 'cart', 'no-user' );
}

const successUrl = thankYouUrl.startsWith( 'http' ) ? thankYouUrl : currentUrl.origin + thankYouUrl;
const cancelUrl = currentUrl.toString();
```

Then, we don't need to parse URL without query and check the length of `updatedQueryString`

Reply:
We just have to be careful that we _do_ strip out the URL hash within the string (added in https://github.com/Automattic/wp-calypso/pull/51653) because it may break nested URLs.

Reply:
> // Actaully, I'm not sure is there any case to cause an error when accessing the `window.location.href` 🤔

It used to break in testing environments since `window` was not available, but the current test runner does seem to set `window` automatically so maybe this guard is unnecessary. Still, I think I'll leave it for now just to be safe.